[Move Replica] Enable StatsManager to dynamically add/remove store

jsjtzyy opened the pull request:
Adding two methods in StatsManager to enable dynamic add/remove store.

xunyin8 commented on ambry-server/src/main/java/com.github.ambry.server/StatsManager.java:
```diff
@@ -166,6 +165,28 @@ StatsSnapshot fetchSnapshot(PartitionId partitionId, List<String> unreachableSto
+   * Remove {@link com.github.ambry.store.BlobStore} from StatsManager and thus it won't be checked during next round of aggregation.
+   * @param id the {@link ReplicaId} associated with store to be removed
+   */
+  void removeBlobStore(ReplicaId id) {
```
Summary from offline discussion: consider the possibility to check the partitionToReplicaMap at the end of stats collection to remove any unreachable replica that no longer exist in the map. This can help with the scenario when a store is removed while stats collection is running and reporting it as unreachable when it's actually being removed.

jsjtzyy replied:
I need to correct the statement I made during offline discussion.  I misunderstood the description in Oracle docs:
```
Iterators and Enumerations return elements reflecting the state of the hash table at some point at or since the creation of the iterator/enumeration
```
This doesn't mean the iterator is an immutable view of map. Based on my experiments, the iterator would access new added element in concurrent case. So, I took @vgkholla's advice and created new Set as "snapshot" of map's keySet(). The relevant results are presented in concurrent unit tests.

jsjtzyy commented:
Addressed comments from @vgkholla and @xunyin8. Added concurrent tests for dynamically adding/removing replica in StatsManager.
./gradlew build && ./gradlew test succeeded.

jsjtzyy commented:
@cgtz reminder to review 
./gradlew build && ./gradlew test succeeded.

jsjtzyy commented:
Addressed @cgtz 's comments.

jsjtzyy commented on ambry-server/src/main/java/com.github.ambry.server/StatsManager.java:
```diff
@@ -166,12 +169,38 @@ StatsSnapshot fetchSnapshot(PartitionId partitionId, List<String> unreachableSto
+   * Remove {@link ReplicaId} from StatsManager and thus it won't be checked during next round of aggregation.
+   * @param id the {@link ReplicaId} associated with store to be removed
+   */
+  void removeReplica(ReplicaId id) {
```
Changed.

jsjtzyy commented on ambry-server/src/main/java/com.github.ambry.server/StatsManager.java:
```diff
@@ -184,27 +213,28 @@ String getNodeStatsInJSON(StatsReportType statsReportType) {
-      List<String> unreachableStores = new ArrayList<>();
-      Iterator<PartitionId> iterator = partitionToReplicaMap.keySet().iterator();
+      List<PartitionId> unreachablePartitions = new ArrayList<>();
+      Iterator<PartitionId> iterator = (new HashSet<>(partitionToReplicaMap.keySet())).iterator();
```
Right, the intention is to get the snapshot of previous keyset.

jsjtzyy commented on ambry-server/src/main/java/com.github.ambry.server/StatsManager.java:
```diff
@@ -184,27 +213,28 @@ String getNodeStatsInJSON(StatsReportType statsReportType) {
+      List<String> unreachableStores = examineUnreachablePartitions(unreachablePartitions);
       metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
       StatsHeader statsHeader = new StatsHeader(StatsHeader.StatsDescription.QUOTA, time.milliseconds(),
           partitionToReplicaMap.keySet().size(), partitionToReplicaMap.keySet().size() - unreachableStores.size(),
```
Thanks for catching this. Fixed the code as you suggest.
